### PR TITLE
[Merged by Bors] - Remove proposals DB

### LINKED ISSUE #5261: remove database access from hare when selecting proposals
image 

preround latency should be about the rest rounds value, maybe slightly more but x3. preround is blue on the grafana screenshot above.
major contributing factor to increased latency is the need to scan proposals and do several checks to determine if they should be upvoted.

see code in https://github.com/spacemeshos/go-spacemesh/blob/develop/hare3/hare.go#L506-L608

it is all in Hare.proposals method. sorry i can't find how to lookup code at selected commit

### PR DESCRIPTION
**Pull request by poszu**
## Motivation
Closes #5261 

## Description
Exchanged proposals SQL table for an in-memory store. It is debatable whether storing proposals in a central place (the `Store`) is required. Proposals could be kept in Hare only and the selected proposals could be passed down to the block generator after Hare completes. However, the set of proposal IDs in Hare's result might contain proposals unavailable locally. In such a scenario, the current code fetches the missing proposals. The blob store must be able to retrieve the requested proposals for the fetcher to work. There is an alternative way - we could remove support for fetching proposals at all and fail block generation in the described scenario - a thing to consider in the future.

#### Changes:
- removed the `proposals` table from DB
- added the in-memory store for proposals (used in `Hare, the ` block `Generator` and the `BlobStore`)
- added a non-exported `beacon` value in `Ballot struct` to cache the beacon value a ballot uses. It's set to either `EpochData.Ballot` or `RefBallot->EpochData.Ballot`
- proposals handler figures out the beacon for a proposal (ballot) and passes the proposal down to Hare
- added a `Has()` method to the `BlobStore` to avoid costly deserialization when not necessary

## Test Plan
Adjusted some unit tests and added new ones

### CODE REVIEW

**Comment by poszu** [comment]
bors try

**Comment by poszu** [comment]
bors try

**Comment by poszu** [comment]
Flake in TestPartition_30_70 (failed to spend)
bors try

File: hare3/hare.go [hunk 1]
```diff
@@ -503,26 +506,15 @@ func (h *Hare) onOutput(session *session, ir IterRound, out output) error {
 	return nil
 }
 
-func (h *Hare) proposals(session *session) []types.ProposalID {
+func (h *Hare) selectProposals(session *session) []types.ProposalID {
```
Comment (dshulyak) [inline hare3/hare.go:509]: i think selecting collection of proposals is still not ideal. 
we can do the same work incrementally, every time when we learned about new proposal.

should not be a problem though as time is likely can't be larger then milliseconds even if number of proposals grows above 1000

File: node/node.go [hunk 2]
```diff
@@ -815,8 +795,17 @@ func (app *App) initServices(ctx context.Context) error {
 	)
 	app.certifier.Register(app.edSgn)
 
+	var evictedLayer types.LayerID
+	if lid := app.clock.CurrentLayer(); lid > 0 {
+		evictedLayer = lid - 1
+	}
+	proposalsStore := store.New(
+		store.WithEvictedLayer(evictedLayer),
```
Comment (dshulyak) [inline node/node.go]: i think it is safe to initialize it to current layer, so that store keeps track only of layers after the one node was started

File: node/node.go [hunk 3]
```diff
@@ -815,8 +795,17 @@ func (app *App) initServices(ctx context.Context) error {
...
+		store.WithStoreLogger(app.addLogger(ProposalStoreLogger, lg).Zap()),
+	)
```
Comment (dshulyak) [inline node/node.go:802]: also we should keep them for the period while hare might be still running, this is referred as zdist in tortoise. capacity should be zdist + 1

File: proposals/store/store.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}
+}
+
+func WithEvictedLayer(layer types.LayerID) StoreOption {
+	return func(s *Store) {
+		s.evicted = &layer
+	}
+}
+
+func WithStoreLogger(logger *zap.Logger) StoreOption {
```
Comment (dshulyak) [inline proposals/store/store.go]: maybe just WithLogger, Store in the name seems redundant

File: proposals/store/store.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	s.logger.Info("proposals store created", zap.Uint32("capacity", s.capacity.Uint32()), zap.Uint32("evicted", evicted))
+	return s
+}
+
+func (s *Store) IsEvicted(layer types.LayerID) bool {
+	return s.evicted != nil && *s.evicted >= layer
+}
+
+func (s *Store) OnLayer(layer types.LayerID) {
```
Comment (dshulyak) [inline proposals/store/store.go:79]: could you please add metric for the number of proposals that we keep in this store, labeled by layer?

File: proposals/store/store.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	return false
+}
+
+func (s *Store) GetBlob(id types.ProposalID) ([]byte, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for layer := range s.data {
+		if proposal, ok := s.data[layer][id]; ok {
+			blob, err := codec.Encode(proposal)
```
Comment (dshulyak) [inline proposals/store/store.go]: it is better to encode without keeping an RLock

File: proposals/handler.go [hunk 7]
```diff
@@ -325,19 +330,11 @@ func (h *Handler) handleProposal(ctx context.Context, expHash types.Hash32, peer
...
-	} else if has {
+	if h.proposals.IsKnown(p.Layer, p.ID()) {
 		known.Inc()
 		return fmt.Errorf("%w proposal %s", errKnownProposal, p.ID())
 	}
-	proposalDuration.WithLabelValues(dbLookup).Observe(float64(time.Since(t1)))
 
-	logger.With().Info("new proposal",
-		log.String("exp hash", expHash.ShortString()),
-		log.Int("num_txs", len(p.TxIDs)))
```
Comment (dshulyak) [inline proposals/handler.go:340]: do we log it somewhere else? why to remove?

File: common/types/ballot.go [hunk 8]
```diff
@@ -68,6 +68,8 @@ type Ballot struct {
 	ballotID BallotID
 	// malicious is set to true if smesher that produced this ballot is known to be malicious.
 	malicious bool
+	// beacon that this ballot references either via EpochData or RefBallot
+	beacon Beacon
```
Comment (dshulyak) [inline common/types/ballot.go]: i think it is used only with proposal. i think we should set this value on *Proposals rather, then ballot

File: fetch/fetch.go [hunk 9]
```diff
@@ -499,7 +501,7 @@ func (f *Fetch) failAfterRetry(hash types.Hash32) {
 	}
 
 	// first check if we have it locally from gossips
-	if _, err := f.bs.Get(req.hint, hash.Bytes()); err == nil {
+	if has, err := f.bs.Has(req.hint, hash.Bytes()); err == nil && has {
```
Comment (dshulyak) [inline fetch/fetch.go]: i agree that it is much better, do you mind submitting this change separately?

File: hare3/protocol.go [hunk 10]
```diff
@@ -482,7 +482,9 @@ func thresholdTallies[T interface {
 	comparable
 	fmt.Stringer
 }](
-	state map[messageKey]*gossipInput, filter IterRound, msgGrade grade,
+	state map[messageKey]*gossipInput,
+	filter IterRound,
+	msgGrade grade,
```
Comment (dshulyak) [inline hare3/protocol.go:487]: why did you change this? i think it passes linter

i don't mind that the formatting change, but it makes it harder to focus on core of the change

File: proposals/handler.go [hunk 11]
```diff
@@ -356,19 +353,17 @@ func (h *Handler) handleProposal(ctx context.Context, expHash types.Hash32, peer
...
-		logger.With().Error("failed to save proposal", log.Err(err))
-		return fmt.Errorf("save proposal: %w", err)
+
+	err = h.proposals.OnProposal(&p)
+	switch {
+	case errors.Is(err, store.ErrProposalExists):
+		known.Inc()
+		return fmt.Errorf("%w %s", errKnownProposal, p.ID())
+	case err != nil:
+		return fmt.Errorf("saving proposal: %w", err)
```
Comment (dshulyak) [inline proposals/handler.go:367]: could you please doublecheck that in both of this cases p2p handler will return Ignore and not Reject. Reject will disconnect a peer, and we don't want that if proposal was evicted or known

File: proposals/handler.go [hunk 12]
```diff
@@ -404,6 +399,25 @@ func (h *Handler) handleProposal(ctx context.Context, expHash types.Hash32, peer
...
+		return fmt.Errorf("empty refballot")
+	}
+	refBallot, err := ballots.Get(h.db, b.RefBallot)
+	if err != nil {
+		return fmt.Errorf("cannot find refballot '%s' in DB: %w", b.RefBallot.String(), err)
+	}
+	if refBallot.EpochData == nil {
+		return fmt.Errorf("refballot '%s' with empty epoch data", b.RefBallot.String())
+	}
+	b.SetBeacon(refBallot.EpochData.Beacon)
```
Comment (dshulyak) [inline proposals/handler.go]: makes more sense if it is kept on proposal

**Review by dshulyak — COMMENTED** [review]
looks good, i left comments abou edge cases. i also need to remind myself how it works with transactions, will do one more review later

File: proposals/handler.go (hunk 7 above)
Comment (poszu) [reply proposals/handler.go:340]: It's logged in `Hare::OnProposal()`

File: hare3/protocol.go (hunk 10 above)
Comment (poszu) [reply hare3/protocol.go:487]: I think I was going through the hare code to understand it and I reformatted it to be able to read it and then staged the change brainlessly.

File: proposals/handler.go (hunk 11 above)
Comment (poszu) [reply proposals/handler.go:367]: Perhaps we should refactor (in a separate effort) the handlers a bit to not use errors for control flow? It could return 2 things:
- a "decision" like `Ok|Ignore|Reject`,
- the error

File: fetch/fetch.go (hunk 9 above)
Comment (poszu) [reply fetch/fetch.go]: I opened a separate PR: https://github.com/spacemeshos/go-spacemesh/pull/5602

File: proposals/handler.go (hunk 11 above)
Comment (poszu) [reply proposals/handler.go:367]: I checked - the p2p `Reject` is returned only if the error is `pubsub.ErrValidationReject`. Any other error results in `Ignore`. https://github.com/spacemeshos/go-spacemesh/blob/61cf5b72667bbf4381e5bb25bba66cf9ef5e49cd/p2p/pubsub/wrapper.go#L44-L51.

File: common/types/ballot.go (hunk 8 above)
Comment (poszu) [reply common/types/ballot.go]: Moved.

**Review by dshulyak — APPROVED** [review]
there is also this table `proposal_transactions`. it is used to track which transactions are referenced in proposals. so for example if block wasn't built yet, but we started next layer we will not reference those transactions again

it technically doesn't need to store a reference to proposal itself, it is sufficient to know at what layer that transaction was included. but it also doesn't hurt to have that

File: proposals/handler.go (hunk 7 above)
Comment (dshulyak) [reply proposals/handler.go:340]: one thing is that it is logged with Debug now, i think it is better to keep it with Info in the handler itself

**Comment by poszu** [comment]
bors merge

File: proposals/handler.go (hunk 7 above)
Comment (poszu) [reply proposals/handler.go:340]: Sorry, I missed your comment. I created a PR with it:  #5610
